swap/src/asb/event_loop.rs

```diff
@@ -318,16 +318,30 @@ where
         &mut self,
         min_buy: bitcoin::Amount,
         max_buy: bitcoin::Amount,
+        monero_wallet: Arc<monero::Wallet>,
     ) -> Result<BidQuote> {
+        let max_bitcoin_for_monero: bitcoin::Amount;
+        let mut max_bitcoin_buy = max_buy;
+        let monero_balance = monero_wallet.get_balance().await?.as_monero();
+
         let rate = self
             .latest_rate
             .latest_rate()
             .context("Failed to get latest rate")?;
 
+        max_bitcoin_for_monero = bitcoin::Amount::from_btc(monero_balance * rate.ask()?.as_btc())?;
+        if max_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                    "Your Max bitcoin to sell: {} is smaller than your current balance of monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
+            max_bitcoin_buy = max_bitcoin_for_monero
+        }
```

Comment:
Please implement this without mutable variables, this is too hard to follow in what is being modified under which conditions.